catalog/app/containers/Bucket/PackageUpdateDialog.js

```diff
@@ -0,0 +1,933 @@
+    marginTop: t.spacing(1),
+  },
+}))
+
+function FilesInput({ input, meta, uploads, setUploads, errors = {} }) {
+  const classes = useFilesInputStyles()
+
+  const value = input.value || { added: {}, deleted: {}, existing: {} }
+  const disabled = meta.submitting || meta.submitSucceeded
+  const error = meta.submitFailed && meta.error
+
+  const onInputChange = input.onChange
+
+  const totalSize = React.useMemo(
+    () => Object.values(value.added).reduce((sum, f) => sum + f.size, 0),
+    [value.added],
+  )
+
+  const warn = totalSize > PD.MAX_SIZE
+
+  // eslint-disable-next-line no-nested-ternary
+  const label = error
+    ? errors[error] || error
+    : warn
+    ? `Total size of new files exceeds recommended maximum of ${readableBytes(
+        PD.MAX_SIZE,
+      )}`
+    : 'Drop files here or click to browse'
+
+  const pipeValue = useMemoEq([onInputChange, value], () => (...fns) =>
+    R.pipe(...fns, onInputChange)(value),
+  )
+
+  const onDrop = React.useCallback(
+    (files) => {
+      if (disabled) return
+      pipeValue(
+        R.reduce(
+          (acc, file) => {
+            const path = getNormalizedPath(file)
+            return R.evolve(
+              {
+                added: R.assoc(path, file),
+                deleted: R.dissoc(path),
+              },
+              acc,
+            )
+          },
+          R.__, // eslint-disable-line no-underscore-dangle
+          files,
+        ),
+      )
+    },
+    [disabled, pipeValue],
+  )
+
+  const rm = (path) => {
+    pipeValue(R.evolve({ added: R.dissoc(path) }))
+    setUploads(R.dissoc(path))
+  }
+
+  const del = (path) => {
+    pipeValue(R.evolve({ deleted: R.assoc(path, true) }))
+  }
+
+  const restore = (path) => {
+    pipeValue(R.evolve({ deleted: R.dissoc(path) }))
+  }
+
+  const handleAction = (handler, ...args) => (e) => {
+    e.stopPropagation()
+    if (disabled) return
+    handler(...args)
+  }
+
+  const revertFiles = React.useCallback(() => {
+    if (disabled) return
+    onInputChange(meta.initial)
+    setUploads({})
+  }, [disabled, onInputChange, setUploads, meta.initial])
+
+  const { getRootProps, getInputProps, isDragActive } = useDropzone({ onDrop })
+
+  const totalProgress = React.useMemo(() => getTotalProgress(uploads), [uploads])
+
+  const computedEntries = useMemoEq(value, ({ added, deleted, existing }) => {
+    const p1 = Object.entries(existing).map(([path, { size }]) => {
+      if (path in deleted) {
+        return { type: 'deleted', path, size }
+      }
+      if (path in added) {
+        const a = added[path]
+        return { type: 'modified', path, size: a.size, delta: a.size - size }
+      }
+      return { type: 'unchanged', path, size }
+    })
+    const p2 = Object.entries(added).reduce((acc, [path, { size }]) => {
+      if (path in existing) return acc
+      return acc.concat({ type: 'added', path, size })
+    }, [])
+    return R.sortBy((x) => [TYPE_ORDER.indexOf(x.type), x.path], p1.concat(p2))
```

Comment:
Shouldn't `sortBy` use a function that returns a number, not an array?

Reply:
`[1, 2] < [1, 2]` works, but I think an explicit number is more comprehensible and less error prone

Reply:
well, we cant convert a string (`x.path`) to a number, and seems like array comparison works fine in js, e.g.
`[1, 'aaa'] < [1, 'aab']`

Reply:
an alternative would be to use `sortWith`:
```javascript
R.sortWith([
  R.ascend(R.pipe(R.prop('type'), R.indexOf(R.__, TYPE_ORDER))),
  R.ascend(R.prop('path'))
])
```
but that seems more verbose and less readable